WcaOnRails/lib/relations.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Relations
+  def self.get_chain(wca_id1, wca_id2)
+    find_chain([[wca_id1]], [[wca_id2]])
+  end
+
+  def self.competitions_together(wca_id1, wca_id2)
+    PeoplePairWithCompetition.includes(:competition).where(wca_id1: wca_id1, wca_id2: wca_id2).map(&:competition)
+  end
+
+  # Internal
+
+  def self.find_chain(left_chains, right_chains)
+    extended_chains_by_one_degree!(left_chains)
+    final_chain = random_final_chain(left_chains, right_chains)
+    return final_chain if final_chain
+    extended_chains_by_one_degree!(right_chains)
+    final_chain = random_final_chain(left_chains, right_chains)
+    return final_chain if final_chain
+    find_chain(left_chains, right_chains)
+  end
+
+  def self.extended_chains_by_one_degree!(chains)
+    chains.map! do |chain|
+      linkings[chain.last].map { |wca_id| [*chain, wca_id] }
+    end.flatten!(1)
+  end
+
+  def self.random_final_chain(left_chains, right_chains)
+    for *left_chain, left_last in left_chains.shuffle
+      for *right_chain, right_last in right_chains.shuffle
+        return [*left_chain, left_last, *right_chain.reverse] if left_last == right_last
+      end
+    end
+    nil
+  end
+
+  def self.linkings
+    @@linkings ||= ActiveRecord::Base.connection.execute("SELECT wca_id, wca_ids FROM linkings")
+                                     .to_a.map! { |wca_id, wca_ids| [wca_id, wca_ids.split(',')] }
+                                     .to_h.freeze
```

Comment:
Actually I don't see any option here. But after trying out I think it would be __much__ better to use sql queries, even if we end up with a bunch of them.

I tried a 4th degree relation, where there were 300 queries, but as we have index on `wca_id` each one run in ~0.3ms and overall SQL really took less than 0.5s.
Actually bigger degrees are so scary, because if two people are loosely-related, then it's likely that they don't have that many related wca-ids.
I found a relation of 3rd degree (between me - I have a lot of related peoples - and some random guy) and it resulted in 1670 queries. That seems terrible but SQL queries for this request took 730ms (all queries, not only to the `linkings` table).

So this sounds like a better alternative (and probably the only one able to run in production without overhead at the moment), as I was becoming scared of our server having to keep ~1GB data, which wouldn't be worth this funny statistics thing.

So basically the question is whether we are ok with numerous queries, considering that they are incredibly fast. I don't see why even 1600 queries running summarily within 0.5s would be worse than any single query running for let's say 1s (we surely have some such, don't we?) =)